packages/@uppy/aws-s3/src/index.js

```diff
@@ -185,105 +208,96 @@ module.exports = class AwsS3 extends Plugin {
+            warnedSuccessActionStatus = true
           }
-
-          // Trim the query string because it's going to be a bunch of presign
-          // parameters for a PUT request—doing a GET request with those will
-          // always result in an error
-          return { location: xhr.responseURL.replace(/\?.*$/, '') }
+          // The responseURL won't contain the object key. Give up.
+          return { location: null }
         }
 
-        return {
-          // Some S3 alternatives do not reply with an absolute URL.
-          // Eg DigitalOcean Spaces uses /$bucketName/xyz
-          location: resolveUrl(xhr.responseURL, getXmlValue(content, 'Location')),
-          bucket: getXmlValue(content, 'Bucket'),
-          key: getXmlValue(content, 'Key'),
-          etag: getXmlValue(content, 'ETag')
-        }
-      },
-
-      // Get the error data from a failed XMLHttpRequest instance.
-      // `content` is the S3 response as a string.
-      // `xhr` is the XMLHttpRequest instance.
-      getResponseError (content, xhr) {
-        // If no response, we don't have a specific error message, use the default.
-        if (!isXml(content, xhr)) {
-          return
+        // responseURL is not available in older browsers.
+        if (!xhr.responseURL) {
+          return { location: null }
         }
-        const error = getXmlValue(content, 'Message')
-        return new Error(error)
+
+        // Trim the query string because it's going to be a bunch of presign
+        // parameters for a PUT request—doing a GET request with those will
+        // always result in an error
+        return { location: xhr.responseURL.replace(/\?.*$/, '') }
+      }
+
+      return {
+        // Some S3 alternatives do not reply with an absolute URL.
+        // Eg DigitalOcean Spaces uses /$bucketName/xyz
+        location: resolveUrl(xhr.responseURL, getXmlValue(content, 'Location')),
+        bucket: getXmlValue(content, 'Bucket'),
+        key: getXmlValue(content, 'Key'),
+        etag: getXmlValue(content, 'ETag')
       }
     }
 
-    // Replace getResponseData() with overwritten version.
-    if (this.opts.getResponseData) {
-      xhrUploadOpts.getResponseData = this.opts.getResponseData
+    // Get the error data from a failed XMLHttpRequest instance.
+    // `content` is the S3 response as a string.
+    // `xhr` is the XMLHttpRequest instance.
+    function defaultGetResponseError (content, xhr) {
+      // If no response, we don't have a specific error message, use the default.
+      if (!isXml(content, xhr)) {
+        return
+      }
+      const error = getXmlValue(content, 'Message')
+      return new Error(error)
     }
 
-    this.uppy.use(XHRUpload, xhrUploadOpts)
+    const xhrOptions = {
+      fieldName: 'file',
+      responseUrlFieldName: 'location',
+      timeout: this.opts.timeout,
+      // Share the rate limiting queue with XHRUpload.
+      __queue: this.requests,
```

Comment:
yeah—the hope is that nobody will use it even if they know it exists after digging into the code, and the `__` is intended as a warning sign

Reply:
In a pipeline redesign this would no longer be necessary as we'd probably do rate limiting as part of the pipeline, and plugins would not need to worry about it anymore.